### PR TITLE
mpreal updated; Brutus interface enabled for mpmath.py

### PR DESCRIPTION
**Pull request by GFTwrt**
Hello,

please check and merge brutus interface.cc and interface.py to master to fulfill #155 as mentioned in #579

Actual the mpmath functions do not support units. So I decided to use SI in this functions only. I can add units to mpf-type variables but get errors in next mathematical operations. 

Thank You!
GFTwrt

### CODE REVIEW

File: src/amuse/community/brutus/Makefile [hunk 1]
```diff
@@ -9,9 +9,9 @@ AMUSE_DIR?=../../../..
 # mpfr: http://www.mpfr.org/
 # mpfrc++: http://www.holoborodko.com/pavel/mpfr/
 # Change these paths in this Makefile and in src/makefile
-GMP_LIBS   ?= -L/data1/boekholt/Packages/mpreal_local/lib/ -lgmp  # libgmp.a
-MPFR_LIBS  ?= -L/data1/boekholt/Packages/mpreal_local/lib/ -lmpfr # libmpfr.a
-MPFR_FLAGS ?= -I/data1/boekholt/Packages/mpreal_local/include/    # mpreal.h
+GMP_LIBS   = -L/usr/lib/x86_64-linux-gnu/ -lgmp  # libgmp.a
+MPFR_LIBS  = -L/usr/lib/x86_64-linux-gnu/ -lmpfr # libmpfr.a
+MPFR_FLAGS = -I./mpfrc++    # mpreal.h
```
Comment (ipelupessy) [inline src/amuse/community/brutus/Makefile]: these should be picked up by the amuse config, so overriding those can break the compilation - why was this necessary?

File: src/amuse/community/brutus/interface.py [hunk 2]
```diff
@@ -182,12 +318,131 @@ def __init__(self, convert_nbody = None, **options):
...
     def initialize_code(self):
         result = self.overridden().initialize_code()
         self.parameters.brutus_output_directory = self.output_directory
         return result
     
+    def get_potential_energy_p_si(self):
+        self.adjust_prec()
+        a=self.convert_nbody.to_si(nbody_system.energy).number
+        b=mpmath.mpf(self.get_potential_energy_string())*a
+        return b #* self.convert_nbody.to_si(nbody_system.energy).unit
```
Comment (ipelupessy) [inline src/amuse/community/brutus/interface.py]: I would put this, and the other get_.._p_si in brutusInterface instead; then they should not need the nbody conversion (can be named get_.._mpreal instead)

Comment (ipelupessy) [reply src/amuse/community/brutus/interface.py]: also: why is the adjust_precision necessary?

**Comment by ipelupessy** [comment]
@tjardaboekholt can you also comment?

File: src/amuse/community/brutus/Makefile (hunk 1 above)
Comment (GFTwrt) [reply src/amuse/community/brutus/Makefile]: My request was to use use the interface.cc and Interface.py. The makefile is not of interest. changes in the makefile was to get it work on my machine. But need not be relevant for the master. I did not know that I can't choose the files for pull request sepaerartely.

Comment (ipelupessy) [reply src/amuse/community/brutus/Makefile]: ok, no problem will fix it..
I don't think you can select files from a pull request, usually I make sure I only commit relevant files in a seperate branch, eg:
```
git checkout master
git branch brutus_mpreal
git checkout brutus_mpreal
git add <file1> <file2> etc
git commit
git push <your remote> brutus_mpreal:brutus_mpreal
```
(a seperate branch is handy since it will allow you to work on multiple things at the same time)

File: src/amuse/community/brutus/interface.py (hunk 2 above)
Comment (GFTwrt) [reply src/amuse/community/brutus/interface.py]: Hello,
you have recognized "import mpmath"? The actual Brutusinterface is not able to handle mpmath. If you want to integrate a complete mpmath handling to your interface I'd welcome it. But for the moment I can't do this. Therefore I put the conversion to the driver to not impact all parts of the system. The adjustment is nessesary to get same precision on both sides of the interface ( Python-mpmath; C++-mpreal)

Comment (ipelupessy) [reply src/amuse/community/brutus/interface.py]: yea, I saw it...I think its ok for now..I am not a 100% sure how mpmath manages the precision in its context..it could also be put in the setter of the precision of Brutus and/or commit_parameters..but don't worry about that..

**Comment by GFTwrt** [comment]
The goal of this is to get the same precision on code and python side. The actual numbers are only 64-bit-floats. But this is not enough for my application. I need at least 50 significant figures. Brutus itself can handle this but the interface was not able to do so. So the only way is to convert mpreal to string and send it as characters trough the interface. At the python side mpmath collects the characters again and converts it to as number again. Mpmath as well as mpmath needs to be configured according to the amount of significant figures.

File: src/amuse/community/brutus/interface.py (hunk 2 above)
Comment (GFTwrt) [reply src/amuse/community/brutus/interface.py]: mpmath allows you to increase and decrease see  http://mpmath.org/doc/current/basics.html?highlight=prec#temporarily-changing-the-precision. But I do just increase. The other direction is not of interest. I think this more easy than handling different setters.

**Comment by tjardaboekholt** [comment]
I did the same as  GFTwrt's last comment, i.e. request the full string from Brutus, and then convert it to mpmath type on the Python side.This in principle should work okay. Alternatively, we can let the interface do this for us.

I found I needed mpmath only if I am interested in the growth of very small perturbations. In other cases where I am interested in the converged solution to the first few (say 6) decimal places, then I do not require all the characters Brutus has, and requesting a double suffices.

**Comment by GFTwrt** [comment]
Hello,
please keep in mind: I haven't deactivated anything. -> The standard interface works like before. But you're able to get the whole precision additionally.

**Comment by ipelupessy** [comment]
ok, I have added some code to make sure everything works if mpmath is not installed - making mpmath a optional dependency..furthermore:
- the PR needs some tests to test the new function (to be added to test/codes_tests/test_brutus.py, it passes current tests so thats ok;-) 0
- id is a python keyword, wherever its used as index argument, it should be replaced by "index"
- i think the _p_si should be moved to BrutusInterface and the unit conversion not in there (thus the name is not ok, should be then .._mpreal (keep in mind you would want to be able to use mpreals also when convert_nbody=None)
- if you want to keep the convenience functions rewrite them to use the lowlevel .._mpreal

**Comment by GFTwrt** [comment]
Hello, didn't you want to merge?

Please check my latest change in the main Makefile to pull too.

**Comment by ipelupessy** [comment]
is it possible to address the 3 issues above (tests and the two regarding the _p_si ? at leasts the tests are very important (the others I can have a look at later, but then its good to have the tests to check nothing breaks...)

**Comment by GFTwrt** [comment]
Can't follow you. What tests are you expecting? This is additional stuff. If the std-automated-Test for brutus works it should be fine. The rest of the interface just delivers a higher count of figures. To bring the things to the normal interface would mean to change the whole interface with all python files to enable "mpmath.py". The naming is p_si and not mpreal, because mpreal is just one implementation of multiprecision (see: https://www.boost.org/doc/libs/1_72_0/libs/multiprecision/doc/html/boost_multiprecision/intro.html  or http://homepages.laas.fr/mmjoldes/campary/ or http://mpmath.org/doc/current/index.html or https://escholarship.org/uc/item/69q5t2mj ...). The "p" is the synonym for "precise". "si" stands for "Système international d’unités". My actual goal is the usage of brutus as a reference for a mp-gadget2 using CAMPARY in BOOST MP-Template evironment to enable my solarwind simulations in our solar system. The precision is needed, because my calculations show that the power of the solar wind is in dimensions of milli Watt, while the total energy is in ranges of 10³⁵ Joule.

**Comment by ipelupessy** [comment]
the new interface functions should be included in the test suite; you can look at test/codes_tests/test_brutus.py, then add a few tests which check that the _p_si methods work (by e.g. setting and retrieving random values)..its to ensure they keep on working in the future if someone works on the brutus interface..

**Comment by GFTwrt** [comment]
test/codes_tests/test_brutus.py just defines classes. Where are they called?

**Comment by ipelupessy** [comment]
normally you call them through nosetest:
```
mpiexec nosetests test/codes_tests/test_brutus.py -sv
```
(-sv gives verbose output) will try to run all tests in the file, you can specify 1 test,  e.g.:
```
mpiexec nosetests test/codes_tests/test_brutus.py:TestBrutusInterface.test1 -sv
```

**Comment by GFTwrt** [comment]
Hello, I'm a little confused about the actual test. I'd interpret it as fail. STDerr looks fine but STDerr is confusing me. Could you please have a look to both channel and comment after.
[stderr.log](https://github.com/amusecode/amuse/files/4372076/stderr.log)
[stdout.log](https://github.com/amusecode/amuse/files/4372079/stdout.log)

**Comment by GFTwrt** [comment]
Just nosetests seems to work.
[just_nosetest.log](https://github.com/amusecode/amuse/files/4372157/just_nosetest.log)

**Comment by GFTwrt** [comment]
Please check test.

**Comment by ipelupessy** [comment]
ok, thanks!
